feat: use xsnap worker CPU meter and start reporting consumption

michaelfig opened the pull request:
Refs #2322

We implement the same basic policy as in the local worker: refill the meter with a constant value (`1e7` computrons) before every crank and terminate the vat if it consumes more than that in a crank.

There is still a long way to go, but at least now we detect and terminate an xsnap vat upon infinite loops.

michaelfig commented on packages/xsnap/src/xsnap.js:
```diff
@@ -73,24 +79,32 @@ export function xsnap(options) {
     stdio: ['ignore', stdout, stderr, 'pipe', 'pipe'],
   });
 
+  let meterStats;
```
I'm not happy with this local state to communicate the meter usage to the `vatCancelled` promise.  It's useful functionality, but ugly.

Any suggestions on how to make this better would be appreciated.

dckc replied:
Perhaps @kriskowal would care to weigh in?

There's a lot going on with async and generators and stuff here. I think I had it all swapped in at one time, but I don't now, and this variable is sufficiently local that I'm OK with it.

I'd prefer to see a type annotation.

michaelfig replied:
I followed @warner's recommendation only to propagate these stats when returning OK (`.`).  It did indeed clear up some of the convolutions.

dckc commented on packages/xsnap/src/xsnap.js:
```diff
@@ -73,24 +79,32 @@ export function xsnap(options) {
   const vatExit = defer();
 
   const args = snapshot ? ['-r', snapshot] : [];
+  if (meteringLimit !== undefined) {
```
for consistency, consider:

```js
const args = [...(snapshot ? ['-r', snapshot] : []),
              ...(meteringLimit ? ['-l', `${meteringLimit}`] : [])]
```

I assume setting `meteringLimit` to 0 is not a useful case.

michaelfig replied:
That's more functional, but I have been criticised in the past for using rest arguments with expressions.

A `meteringLimit` of 0 is exactly how one needs to disable metering, so I've updated the conditional accordingly.

dckc commented on packages/xsnap/src/xsnap.c:
```diff
@@ -167,21 +168,21 @@ static int fxSnapshotWrite(void* stream, void* address, size_t size)
 static xsBooleanValue gxMeteringPrint = 0;
 
-static FILE *fromParent;
-static FILE *toParent;
```
more globals. Sigh. I guess they're "write once" at least.

dckc commented on packages/xsnap/src/xsnap.c:
```diff
@@ -1310,6 +1320,22 @@ static int fxWriteNetString(FILE* outStream, char prefix, char* buf, size_t leng
 
+static int fxWriteAndResetMeterStats(xsMachine *the, FILE* outStream) {
+#if mxMetering
+  char meterIndexBuf[32];
```
again, consistency suggests tabs for indentation

warner replied:
I wouldn't say no to a PR that reindented the entire file with spaces...

michaelfig replied:
> I wouldn't say no to a PR that reindented the entire file with spaces...

I'd suggest creating a `.clang-format` file and enforcing the formatting of our C code.

But that's for the future.  For now, I'm going for consistency.

dckc replied:
> I wouldn't say no to a PR that reindented the entire file with spaces...

Using a style that's inconsistent with Moddable's seems like a bad idea. In particular, it suggests that we're going to be writing / maintaining amounts of code comparable to what they produce, which is not a direction I want to head.

dckc commented on packages/xsnap/src/xsnap.js:
```diff
@@ -106,9 +120,16 @@ export function xsnap(options) {
-   * @returns {Promise<Uint8Array>}
+   * @typedef {Object} RunResult
+   * @property {Uint8Array} reply
+   * @property {any} meterStats
```
is it important to use `any` rather than `unknown`?

dckc commented on packages/xsnap/src/xsnap.js:
```diff
@@ -106,9 +120,16 @@ export function xsnap(options) {
 
   /**
-   * @returns {Promise<Uint8Array>}
+   * @typedef {Object} RunResult
```
`Object` is more or less `any`, yes?

How about:

```
@typedef {{ reply: Uint8Array, meterStats: unknown }} RunResult
```

michaelfig replied:
> `Object` is more or less `any`, yes?

Not when it has any `@property` tags underneath it.

dckc commented on packages/xsnap/src/xsnap.js:
```diff
@@ -194,10 +217,11 @@ export function xsnap(options) {
   /**
    * @param {string} message
-   * @returns {Promise<string>}
+   * @returns {Promise<{ reply: string, meterStats: any }>}
```
again, any particular reason for `any` rather than `unknown`?

michaelfig replied:
Now typed.

dckc commented on packages/SwingSet/src/kernel/vatManager/manager-subprocess-xsnap.js:
```diff
@@ -159,14 +161,20 @@ export function makeXsSubprocessFactory({
+      /*
+      return {
+        reply: result.reply,
+        crankStats: { workerType: 'xsnap-0.0.0', usedMeter: result.meterStats },
```
The purpose of this comment seems to be to document the type of the return. Any particular reason not to use jsdoc / typescript to do that?

michaelfig replied:
It was more of a TODO to describe what I wanted for a return result.

dckc commented on packages/xsnap/makefiles/mac/xsnap.mk:
```diff
@@ -32,6 +32,7 @@ C_OPTIONS = \
 	-DmxRun=1 \
 	-DmxSloppy=1 \
 	-DmxSnapshot=1 \
+	-DmxMetering=1 \
```
https://github.com/Moddable-OpenSource/moddable/issues/550 says the metering code is experimental. I wonder whether we should ask @dtribble or somebody before we merge it to master.

dckc replied:
@dtribble says ok.

dckc reviewed (dismissed):
I'm hesitant to move forward with `-DmxMetering=1` while Moddable considers it experimental. I'd like to hear from @dtribble on that, but presuming you're already in communication with him about it, I approve.

I'd really rather not have `any` in the type annotations, but I don't suppose that's critical.

warner commented on packages/xsnap/src/xsnap.js:
```diff
@@ -121,7 +142,7 @@ export function xsnap(options) {
         throw new Error('xsnap protocol error: received empty message');
       } else if (message[0] === OK) {
-        return message.subarray(1);
+        return { reply: message.subarray(1), meterStats };
```
Wait, shouldn't the `meterStats` be part of the `OK` that comes back from the worker? Why add a separate message for it? Metering should be measured exactly from the beginning to the end of the crank, and `OK` is how the worker signals the end of the crank, so I'm not in favor of a pathway that suggests that the worker might send metering results at other times.

dckc replied:
> Wait, shouldn't the `meterStats` be part of the `OK` that comes back from the worker?

I thought it was, at a higher level... but I only see it in the manager...
https://github.com/Agoric/agoric-sdk/pull/2384/files#diff-73740b92f3f3365578c6c1492419c760b2ec870fa71a269349faca6bfb55f63aR164

I'm not sure how this works without any changes to the supervisor. hmm.

michaelfig replied:
I've improved this now.  Will request a re-review when it's ready.

michaelfig commented on packages/SwingSet/src/kernel/vatManager/manager-subprocess-xsnap.js:
```diff
@@ -135,18 +137,18 @@ export function makeXsSubprocessFactory({
     }
 
-    /** @type { (item: Tagged) => Promise<Tagged> } */
+    /** @type { (item: Tagged) => Promise<{ reply: Tagged, meterStats: any }> } */
```
Fixed.

michaelfig commented on packages/xsnap/src/xsnap.c:
```diff
@@ -802,6 +810,7 @@ void fx_Array_prototype_meter(xsMachine* the)
 
 void fxPatchBuiltIns(txMachine* machine)
 {
+#if 0
```
Done.

dckc commented on packages/xsnap/src/xsnap.js:
```diff
@@ -119,17 +114,43 @@ export function xsnap(options) {
 
+  /** @type {string} */
+  let xsWorkerType;
+  async function getWorkerType() {
```
it's awkward to have the xsnap level know anything about workers.

michaelfig replied:
The worker type is specifically what version of xsnap is in use.

dckc replied:
ah. different sense of "worker"

dckc commented on packages/xsnap/src/xsnap.c:
```diff
@@ -3,6 +3,8 @@
 #include "xsSnapshot.h"
 #include "xs.h"
 
+#define XSNAP_VERSION "0.1.0"
```
how about a test that this matches the package version?

michaelfig replied:
Ah, that's where it's supposed to come from?  I'll `-DXSNAP_VERSION=...` in the Makefiles.

michaelfig commented on packages/SwingSet/src/kernel/vatManager/manager-subprocess-xsnap.js:
```diff
@@ -3,6 +3,9 @@ import { assert, details } from '@agoric/assert';
 import { makeTranscriptManager } from './transcript';
 import { createSyscall } from './syscall';
 
+import '../../types';
```
The only magic is if a JSDoc type is defined in a file that only has comments, that type is available in all files in the same directory.  `VatPowers` was defined in a different directory, so we need to import them to get them in scope.

michaelfig commented on packages/SwingSet/src/kernel/vatManager/manager-subprocess-xsnap.js:
```diff
@@ -26,7 +29,6 @@ const decoder = new TextDecoder();
  * @typedef { ReturnType<typeof import('@agoric/xsnap').xsnap> } XSnap
  * @typedef { ReturnType<typeof import('../state/kernelKeeper').default> } KernelKeeper
  * @typedef { ReturnType<typeof import('./manager-nodeworker').makeNodeWorkerVatManagerFactory> } VatManagerFactory
- * @typedef { [unknown, ...unknown[]] } Tagged
```
> `vatManager/types.js` doesn't seem to be reused anywhere else.

It will be soon.  In the next PR, if I'm lucky.

dckc reviewed (commented):
I think I should withdraw my approval until we have clarity on @warner 's question:
> shouldn't the meterStats be part of the OK that comes back from the worker?

dckc reviewed (commented):
I was about to approve this, but then ava hangs when I `yarn test` because some xsnap process is hanging around.

dckc commented:
> ... ava hangs when I `yarn test` because some xsnap process is hanging around.

diagnosis: `makefiles/lin/xsnap.mk` is missing `-DmxMetering=1`

How did this get thru ci?

eek! `packages/xsnap` doesn't seem to get its tests run in ci: https://github.com/Agoric/agoric-sdk/pull/2384/checks?check_run_id=1875722804

dckc reviewed (changes requested):
`makefiles/lin/xsnap.mk` is missing `-DmxMetering=1`

and the test-quick ci job is missing `packages/xsnap` altogether!

warner commented on packages/xsnap/src/xsnap.c:
```diff
@@ -204,7 +218,7 @@ int main(int argc, char* argv[])
 
 	txSnapshot snapshot = {
-		"xsnap 0.1.0",
+		"xsnap-" XSNAP_VERSION,
```
That's a significant compatibility policy we're baking into the snapshot format right there. It means we're claiming 1: we won't make any snapshot-compatibility-breaking changes without also changing the package.json version, and 2: every package.json version change means a compatibility-breaking change. This sort of binding between "marketing" version and compatibility version means we cannot commit a change to the XS code without simultaneously making a release (otherwise we'd be claiming compatibility for the tree before and after the XS commit).

If we're going to move away from a fixed `0.1.0` now, I'd recommend introducing a separate "snapshot compatibility version" (a sequential integer, not a semver string), instead of tying it to the package.json version. (Storing that snapshot compatibility version under a separate key in `package.json` is a fine idea, assuming it doesn't conflict with anything else that parses the file.. I can't think of a better place to put it).

michaelfig replied:
I've introduced a `SNAPSHOT_SIGNATURE` manifest constant.

warner commented on packages/xsnap/src/xsnap.c:
```diff
@@ -308,10 +322,15 @@ int main(int argc, char* argv[])
 		char done = 0;
 		while (!done) {
+			// By default, use the infinite meter.
+      gxCurrentMeter = 0;
```
as dckc pointed out, this doesn't match the tab/spaces convention of the original, this looks weirdly indented

warner commented on packages/xsnap/src/xsnap.c:
```diff
@@ -880,8 +909,8 @@ void fx_clearTimer(txMachine* the)
 	if (job) {
-        xsForget(job->self);
-        xsSetHostData(xsArg(0), NULL);
+		xsForget(job->self);
```
I think this is a gratuitous whitespace change

michaelfig replied:
It is not.  I'm normalising the spaces in the file to be tabs.

warner commented on packages/xsnap/src/xsnap.js:
```diff
@@ -121,7 +163,25 @@ export function xsnap(options) {
         throw new Error('xsnap protocol error: received empty message');
       } else if (message[0] === OK) {
-        return message.subarray(1);
+        let compute = NaN;
```
`NaN` is a pain to deal with. In this case, it means "unknown", right? As in, the child didn't tell us? Could it be simpler if we require the child to always tell us? And/or use `undefined` to mean "unknown"?

michaelfig replied:
> And/or use `undefined` to mean "unknown"?

I will use `null` since that is JSON-serializable.

warner commented on packages/xsnap/src/xsnap.js:
```diff
@@ -121,7 +163,25 @@ export function xsnap(options) {
+        let compute = NaN;
+        const semi = message.indexOf(SEMI, 1);
+        if (semi >= 0) {
+          // The message is `.938586;`.
```
I think I understand the need for the semicolon encoding now, but it feels off. Fairly soon, we're going to be returning multiple values from the child, right? Not just CPU usage, but something about peak/current memory allocation? I suspect we'll need something more flexible than `/\.\d+;/`. I guess the deeper issue is the handoff in `xsnap.c` between the JS supervisor code (stashing some "how did it go" value, a string I guess) and the C code that needs to wrap that in a netstring and send it back over the pipe. On one hand, I want the supervisor code to get the metering values, so it can include them next to any status data before it `JSON.stringify`s them into the stashed place. On the other hand, I want the meter to be sampled *after* all JS has finished running, so we aren't looking at meter values while the meter is still in use (which additionally raises the question of what happens if the meter expires after we've sampled it but before we return back into C).

I guess I'm avoiding something: cranks failing for metering reasons is very different than cranks failing for e.g. illegal syscalls. The (JS) supervisor can discover (and report) the latter. But a metering failure will exit the entire JS engine, including the supervisor, so both a metering failure and reporting the meter remaining are things that the C code must do, which means they're necessarily delivered in a different chunk of data than whatever the supervisor provides. Bummer.

Ok so maybe the generalized answer is for each parent-to-child message to result in *two* netstrings in response, glued together. The first can be what the C code reports ("success: no metering violation, here's the remaining meter, or failure because of meter underflow), and the second can be what the JS supervisor code reports ("success: no illegal syscalls"). If the C code sees an error, the second value may be empty (perhaps we synthesize an empty object, just `2:{},`.

michaelfig replied:
> Ok so maybe the generalized answer is for each parent-to-child message to result in two netstrings in response, glued together.

I had implemented something like that in the first version of this PR (with the `'M'` command code to report metering results), but it was shot down.  Do you want me to resurrect it?

michaelfig replied:
> I suspect we'll need something more flexible than `/\.\d+;/`

There's nothing preventing the evolution of the coupling between `xsnap.c` and `xsnap.js`.  I half-considered `/\.[^\0]*\0/` to encode arbitrary JSON, but opted for the minimal starting point.

warner reviewed (approved):
Tentative approval, please check the comments and see if the recommended changes are ones you're comfortable with. The XSNAP_VERSION thing is the one I'm most concerned about. The message-result protocol is the one I thought the most about but it might not be worth doing anything about.

michaelfig commented on packages/SwingSet/src/kernel/vatManager/manager-subprocess-xsnap.js:
```diff
@@ -135,18 +137,18 @@ export function makeXsSubprocessFactory({
       parentLog(item[0], '...', item.length - 1);
-      const txt = await worker.issueStringCommand(JSON.stringify(item));
-      const reply = JSON.parse(txt);
+      const result = await worker.issueStringCommand(JSON.stringify(item));
```
Yes, this PR is a baby step.

michaelfig commented on packages/xsnap/src/netstring.js:
```diff
@@ -72,6 +72,7 @@ export async function* reader(input, name = '<unknown>', capacity = 1024) {
       `Unexpected dangling message at offset ${offset} of ${name}`,
     );
   }
+  return new Uint8Array(0);
```
For a JS generator, the `yield X` results in `{ done: false, value: X }` and `return X` results in `{ done: true, value: Y }`.  The value `Y` is not a signal, it's just an empty value with the same type as `X` so that Typescript doesn't get confused.

michaelfig commented on packages/xsnap/src/xsnap.js:
```diff
@@ -14,10 +14,15 @@ import { defer } from './defer';
 import * as node from './node-stream';
 
+// This will need adjustment, but seems to be fine for a start.
+const DEFAULT_CRANK_METERING_LIMIT = 1e7;
```
Keepers and more flexible metering will need a lot more effort than just passing a meter value into every evaluation.  I don't want to overdesign something we'll have to change again.

michaelfig commented on packages/xsnap/src/xsnap.js:
```diff
@@ -105,8 +115,40 @@ export function xsnap(options) {
+    await new Promise(resolve => {
+      // Inititialize the worker type before sending messages.
+      const bufs = [];
+      const xsnapQueryVersion = spawn(bin, ['-v'], {
```
I've ripped it out in favour of a manifest constant, `METER_TYPE`, which is initially `'xs-meter-1'`.

michaelfig commented:
> > ... ava hangs when I `yarn test` because some xsnap process is hanging around.
> 
> diagnosis: `makefiles/lin/xsnap.mk` is missing `-DmxMetering=1`

I've added preflight checks to xsnap.c to reject metering unless `-DmxMetering`.

> How did this get thru ci?
> 
> eek! `packages/xsnap` doesn't seem to get its tests run in ci: https://github.com/Agoric/agoric-sdk/pull/2384/checks?check_run_id=1875722804

Added!
